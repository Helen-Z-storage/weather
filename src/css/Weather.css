
.single_city {
  border-bottom: 1px solid #EFEFEF;
  position: relative;
}

.single_city li{
  padding-left: 20px;
  margin: 0px;
}