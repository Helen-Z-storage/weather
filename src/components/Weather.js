import React, {useState} from 'react';
import '../css/Weather.css';
import WeatherData from './WeatherData';
import Expand from './Expand';
import Tag from './Tag';

const storage = window.localStorage;

const tag_spliter = ",";
// adding valid tags in localStorage's current student data
const addNewTag = (filter, keyCode, new_tags) => {
    let clear = false;
    let message = "";

    // handle case of tag_spliter exist in input value
    if (filter.indexOf(tag_spliter) > -1) {
        message = `No "${tag_spliter}" should include in Tag`;
        clear = true;
    }

    // handle case of invalid empty tag
    else if (keyCode === 13 && filter === "") {
        message = `No empty tag should been add`;
        clear = true;
    }

    // saving tag iff press "enter"
    else if (keyCode === 13) {

        var tag = filter + tag_spliter;

        // Add new tag if tag is not store in current student
        if (new_tags.toUpperCase().indexOf(tag_spliter + tag.toUpperCase()) === -1) {
            new_tags += (tag);
        } else {
            // handle case of duplicate tag
            message = `No duplicate tag should been add`;
        }
        // clear text input field
        clear = true;
    }
    return [clear, new_tags, message];
}

// LocalStorage Hook
function useLocalStorage(cityID, setTagErrorMsg) {
    // State to store our value
    // Pass initial state function to useState so logic is only executed once
    const [prevState, setPrevState] = useState([]);
    const [city, setCity] = useState(() => {
        try {
            // Get from local storage by cityID
            const existCity = storage.getItem(cityID);
            return existCity ? JSON.parse(existCity) : {};
        } catch (error) {
            console.error('Error:', error);
            return {};
        }
    });

    // update student's tag by value of input tagEvent
    const setNewCity = tagEvent => {
        try {
            // adding tag and clear tag adding field by cases
            const [clear, new_tags, tagError] = addNewTag(tagEvent.target.value, tagEvent.nativeEvent.keyCode, city.tags);
            if (clear) {
                tagEvent.target.value = "";
                city.tags = new_tags;
            };
            setTagErrorMsg(tagError);

            // Save new student in state and update new state for synchronize render
            setCity(city);
            setPrevState([...prevState]);

            // Save new student to local storage
            storage.setItem(cityID, JSON.stringify(city));
        } catch (error) {
            console.error('Error:', error);
        }
      };

    return [city, setNewCity];
}

function Weather(props) {
    const {id, expand, cityFilter, countryFilter, tagFilter, 
        handleExpand, handleFilter} = props;

    // init student local storage update hook
    const [tagErrorMsg, setTagErrorMsg] = useState("");
    const [city, setNewCity] = useLocalStorage(id, setTagErrorMsg);

    let removeByCity, removeByCountry, removeByTag, disp;
    removeByCity = city.city.toUpperCase().indexOf(cityFilter) === -1;
    removeByCountry = city.country.toUpperCase().indexOf(countryFilter) === -1;
    removeByTag = city.tags.toUpperCase().indexOf(tagFilter) === -1;

    disp = (removeByCity || removeByCountry || removeByTag)? {display:"none"}: {};

    return (
        <li className="single_city" style={disp}>
            <table>
                <tbody>
                    <WeatherData currCity={city}/>
                    <tr>
                        <td></td>
                        <td>
                            <Expand id={id} currCity={city}
                                expand={expand} handleExpand={handleExpand} />
                            <Tag tags={city.tags} tagErrorMsg={tagErrorMsg} 
                                handleFilter={handleFilter}
                                setNewCity={setNewCity}/>
                        </td>
                    </tr>
                </tbody>
            </table>
        </li>
    );
}

export default Weather;

