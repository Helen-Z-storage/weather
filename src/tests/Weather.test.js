import React from 'react'
import { create } from "react-test-renderer";
import { render, screen } from '@testing-library/react';

import '@testing-library/jest-dom/extend-expect' // expect function
import userEvent from '@testing-library/user-event'; 

import Weather from '../components/Weather';

// required prop data of create component
const tag_spliter = ",";

const id = 0;
const expandT = true;
const expandF = false;
const cityFilter = "";
const filtCityFilter = "KHU";
const unfiltCityFilter = "KHS";
const countryFilter = "";
const filtCountryFilter = "KHA";
const unfiltCountryFilter = "KHB";
const tagFilter = "";
const filtTagFilter = "AD";
const unfiltTagFilter = "AW";
const duplicate_tag = "MENT{enter}";
const non_duplicate_tag = "AD{enter}";
const tag_split_tag = `R%${tag_spliter}`;
const empty_tag = "{enter}";
const handleExpand = jest.fn();
const handleFilter = jest.fn();
const storage = window.localStorage;

// store this city as 0 in mocked localstorage
const city_data = {
    country: "Abkhazia",
    city: "Sukhumi",
    lat: 43.0033629,
    lon: 41.0192741,
    tags: ",adver,tise,Ment,339%?,",
    icon: "https://openweathermap.org/img/wn/03d@2x.png",
    curr_forecast: {
        timestamp: "2023-01-18 09:00:00",
        temp: 11.57,
        feels_like: 10.06,
        temp_min: 11.57,
        temp_max: 14.54,
        icon: "https://openweathermap.org/img/wn/03d@2x.png"
    },
    detail_forecast: [
        {
            timestamp: "2023-01-18 09:00:00",
            temp: 11.57,
            feels_like: 10.06,
            temp_min: 11.57,
            temp_max: 14.54,
            icon: "https://openweathermap.org/img/wn/03d@2x.png"
        },
        {
            timestamp: "2023-01-18 12:00:00",
            temp: 13.2,
            feels_like: 11.83,
            temp_min: 13.2,
            temp_max: 14.75,
            icon: "https://openweathermap.org/img/wn/03d@2x.png"
        },
        {
            timestamp: "2023-01-18 15:00:00",
            temp: 12.26,
            feels_like: 10.95,
            temp_min: 12.26,
            temp_max: 12.26,
            icon: "https://openweathermap.org/img/wn/04n@2x.png"
        },
        {
            timestamp: "2023-01-18 18:00:00",
            temp: 11.37,
            feels_like: 9.97,
            temp_min: 11.37,
            temp_max: 11.37,
            icon: "https://openweathermap.org/img/wn/04n@2x.png"
        },
        {
            timestamp: "2023-01-18 21:00:00",
            temp: 11.01,
            feels_like: 9.55,
            temp_min: 11.01,
            temp_max: 11.01,
            icon: "https://openweathermap.org/img/wn/04n@2x.png"
        },
        {
            timestamp: "2023-01-19 00:00:00",
            temp: 10.28,
            feels_like: 8.7,
            temp_min: 10.28,
            temp_max: 10.28,
            icon: "https://openweathermap.org/img/wn/04n@2x.png"
        },
        {
            timestamp: "2023-01-19 03:00:00",
            temp: 9.95,
            feels_like: 8.9,
            temp_min: 9.95,
            temp_max: 9.95,
            icon: "https://openweathermap.org/img/wn/02n@2x.png"
        },
        {
            timestamp: "2023-01-19 06:00:00",
            temp: 11.08,
            feels_like: 9.47,
            temp_min: 11.08,
            temp_max: 11.08,
            icon: "https://openweathermap.org/img/wn/02d@2x.png"
        }
    ]
};

storage.setItem(id, JSON.stringify(city_data));

describe('testing Weather component with data', () => {
    test("Matches the snapshot: shows expanded Expand component", () => {
        // create component
        const city = create(
            <Weather id={id} expand={expandT}
                     cityFilter={cityFilter} countryFilter={countryFilter} tagFilter={tagFilter}
                     handleExpand={exp_id => handleExpand(exp_id)}
                     handleFilter={(e) => handleFilter(e)}/>
        );

        // expecting output
        expect(city.toJSON()).toMatchSnapshot();
    });

    test("Matches the snapshot: shows non-expanded Expand component", () => {
        // create component
        const city = create(
            <Weather id={id} expand={expandF}
                     cityFilter={cityFilter} countryFilter={countryFilter} tagFilter={tagFilter}
                     handleExpand={exp_id => handleExpand(exp_id)}
                     handleFilter={(e) => handleFilter(e)}/>
        );

        // expecting output
        expect(city.toJSON()).toMatchSnapshot();
    });

    // testing filters
    test("shows filt city displayed city", () => {
        // render component
        render(
            <Weather id={id} expand={expandF}
                     cityFilter={filtCityFilter} countryFilter={countryFilter} tagFilter={tagFilter}
                     handleExpand={exp_id => handleExpand(exp_id)}
                     handleFilter={(e) => handleFilter(e)}/>
        );
        // expecting output
        expect(screen.queryByText("Sukhumi")).toBeVisible();
    });

    test("shows unfilt city displayed city", () => {
        // render component
        render(
            <Weather id={id} expand={expandF}
                     cityFilter={unfiltCityFilter} countryFilter={countryFilter} tagFilter={tagFilter}
                     handleExpand={exp_id => handleExpand(exp_id)}
                     handleFilter={(e) => handleFilter(e)}/>
        );

        // expecting output
        expect(screen.queryByText("Sukhumi")).not.toBeVisible();
    });
    
    // testing filters
    test("shows filt country displayed city", () => {
        // render component
        render(
            <Weather id={id} expand={expandF}
                     cityFilter={cityFilter} countryFilter={filtCountryFilter} tagFilter={tagFilter}
                     handleExpand={exp_id => handleExpand(exp_id)}
                     handleFilter={(e) => handleFilter(e)}/>
        );
        // expecting output
        expect(screen.queryByText("Sukhumi")).toBeVisible();
    });

    test("shows unfilt country displayed city", () => {
        // render component
        render(
            <Weather id={id} expand={expandF}
                     cityFilter={cityFilter} countryFilter={unfiltCountryFilter} tagFilter={tagFilter}
                     handleExpand={exp_id => handleExpand(exp_id)}
                     handleFilter={(e) => handleFilter(e)}/>
        );

        // expecting output
        expect(screen.queryByText("Sukhumi")).not.toBeVisible();
    });

    test("shows filt tag displayed city", () => {
        // render component
        render(
            <Weather id={id} expand={expandF}
                     cityFilter={cityFilter} countryFilter={countryFilter} tagFilter={filtTagFilter}
                     handleExpand={exp_id => handleExpand(exp_id)}
                     handleFilter={(e) => handleFilter(e)}/>
        );

        // expecting output
        expect(screen.queryByText("Sukhumi")).toBeVisible();
    });

    test("shows unfilt tag displayed city", () => {
        // render component
        render(
            <Weather id={id} expand={expandF}
                     cityFilter={cityFilter} countryFilter={countryFilter} tagFilter={unfiltTagFilter}
                     handleExpand={exp_id => handleExpand(exp_id)}
                     handleFilter={(e) => handleFilter(e)}/>
        );

        // expecting output
        expect(screen.queryByText("Sukhumi")).not.toBeVisible();
    });
    
    // test custom hook
    test("tested custom hook useLocalStorage(): adding tag", async () => {
        const user = userEvent.setup();
        // render component
        render(
            <Weather id={id} expand={expandF}
                     cityFilter={cityFilter} countryFilter={countryFilter} tagFilter={tagFilter}
                     handleExpand={exp_id => handleExpand(exp_id)}
                     handleFilter={(e) => handleFilter(e)}/>
        );
        // expecting output
        // adding duplicate tag
        const textfield = screen.getByPlaceholderText("Add a tag");
        await user.type(textfield, duplicate_tag);
        let city = JSON.parse(storage.getItem(id));
        let tags = city_data.tags;
        expect(city.tags).toEqual(tags);
        screen.debug();
        expect(screen.queryByText(`No duplicate tag should been add`)).not.toEqual(null);
        expect(alert.mock.calls[0][0]).toEqual();

        // adding invalid empty tag
        await user.type(textfield, empty_tag);
        city = JSON.parse(storage.getItem(id));
        expect(city.tags).toEqual(tags);
        expect(alert).toBeCalledTimes(2);
        expect(alert.mock.calls[1][0]).toEqual(`No empty tag should been add`);

        // adding invalid tag_split tag
        await user.type(textfield, tag_split_tag);
        city = JSON.parse(storage.getItem(id));
        expect(city.tags).toEqual(tags);
        expect(alert).toBeCalledTimes(3);
        expect(alert.mock.calls[2][0]).toEqual(`No "${tag_spliter}" should include in Tag`);

        // adding non-duplicate tag
        await user.type(textfield, non_duplicate_tag + "{enter}");
        city = JSON.parse(storage.getItem(id));
        expect(city.tags).toEqual(tags + non_duplicate_tag + tag_spliter);
    });
});