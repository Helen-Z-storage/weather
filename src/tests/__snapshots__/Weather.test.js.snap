// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing Weather component with data Matches the snapshot: shows expanded Expand component 1`] = `
<li
  className="single_city"
  style={Object {}}
>
  <table>
    <tbody>
      <tr>
        <td>
          <img
            alt="Sukhumi weather icon"
            className="curr_weather"
            src="https://openweathermap.org/img/wn/03d@2x.png"
          />
        </td>
        <td>
          <ul
            className="data_space"
          >
            <li>
              <div
                className="city_name"
                style={
                  Object {
                    "width": "100%",
                  }
                }
              >
                Sukhumi
              </div>
            </li>
            <li>
              <div
                className="country_name"
                style={
                  Object {
                    "width": "100%",
                  }
                }
              >
                Abkhazia
              </div>
            </li>
          </ul>
        </td>
      </tr>
      <tr>
        <td />
        <td>
          <ul
            className="detailWeather"
          >
            <li>
              <button
                className="expand"
                onClick={[Function]}
              >
                -
              </button>
            </li>
            <li>
              <table>
                <thead>
                  <tr>
                    <td>
                      Date & time
                    </td>
                    <td>
                      Forcast Weather
                    </td>
                    <td>
                      Forcast Temp
                    </td>
                    <td>
                      Forcast Body Temp
                    </td>
                    <td>
                      Forcast Max Temp
                    </td>
                    <td>
                      Forcast Min Temp
                    </td>
                  </tr>
                </thead>
                <tbody>
                  <tr>
                    <td>
                      undefined
                    </td>
                    <td>
                      <img
                        alt="Sukhumi 2023-01-18 09:00:00 weather icon"
                        className="expand_img"
                        src="https://openweathermap.org/img/wn/03d@2x.png"
                      />
                    </td>
                    <td>
                      11.57 °C
                    </td>
                    <td>
                      10.06 °C
                    </td>
                    <td>
                      11.57 °C
                    </td>
                    <td>
                      14.54 °C
                    </td>
                  </tr>
                  <tr>
                    <td>
                      undefined
                    </td>
                    <td>
                      <img
                        alt="Sukhumi 2023-01-18 12:00:00 weather icon"
                        className="expand_img"
                        src="https://openweathermap.org/img/wn/03d@2x.png"
                      />
                    </td>
                    <td>
                      13.2 °C
                    </td>
                    <td>
                      11.83 °C
                    </td>
                    <td>
                      13.2 °C
                    </td>
                    <td>
                      14.75 °C
                    </td>
                  </tr>
                  <tr>
                    <td>
                      undefined
                    </td>
                    <td>
                      <img
                        alt="Sukhumi 2023-01-18 15:00:00 weather icon"
                        className="expand_img"
                        src="https://openweathermap.org/img/wn/04n@2x.png"
                      />
                    </td>
                    <td>
                      12.26 °C
                    </td>
                    <td>
                      10.95 °C
                    </td>
                    <td>
                      12.26 °C
                    </td>
                    <td>
                      12.26 °C
                    </td>
                  </tr>
                  <tr>
                    <td>
                      undefined
                    </td>
                    <td>
                      <img
                        alt="Sukhumi 2023-01-18 18:00:00 weather icon"
                        className="expand_img"
                        src="https://openweathermap.org/img/wn/04n@2x.png"
                      />
                    </td>
                    <td>
                      11.37 °C
                    </td>
                    <td>
                      9.97 °C
                    </td>
                    <td>
                      11.37 °C
                    </td>
                    <td>
                      11.37 °C
                    </td>
                  </tr>
                  <tr>
                    <td>
                      undefined
                    </td>
                    <td>
                      <img
                        alt="Sukhumi 2023-01-18 21:00:00 weather icon"
                        className="expand_img"
                        src="https://openweathermap.org/img/wn/04n@2x.png"
                      />
                    </td>
                    <td>
                      11.01 °C
                    </td>
                    <td>
                      9.55 °C
                    </td>
                    <td>
                      11.01 °C
                    </td>
                    <td>
                      11.01 °C
                    </td>
                  </tr>
                  <tr>
                    <td>
                      undefined
                    </td>
                    <td>
                      <img
                        alt="Sukhumi 2023-01-19 00:00:00 weather icon"
                        className="expand_img"
                        src="https://openweathermap.org/img/wn/04n@2x.png"
                      />
                    </td>
                    <td>
                      10.28 °C
                    </td>
                    <td>
                      8.7 °C
                    </td>
                    <td>
                      10.28 °C
                    </td>
                    <td>
                      10.28 °C
                    </td>
                  </tr>
                  <tr>
                    <td>
                      undefined
                    </td>
                    <td>
                      <img
                        alt="Sukhumi 2023-01-19 03:00:00 weather icon"
                        className="expand_img"
                        src="https://openweathermap.org/img/wn/02n@2x.png"
                      />
                    </td>
                    <td>
                      9.95 °C
                    </td>
                    <td>
                      8.9 °C
                    </td>
                    <td>
                      9.95 °C
                    </td>
                    <td>
                      9.95 °C
                    </td>
                  </tr>
                  <tr>
                    <td>
                      undefined
                    </td>
                    <td>
                      <img
                        alt="Sukhumi 2023-01-19 06:00:00 weather icon"
                        className="expand_img"
                        src="https://openweathermap.org/img/wn/02d@2x.png"
                      />
                    </td>
                    <td>
                      11.08 °C
                    </td>
                    <td>
                      9.47 °C
                    </td>
                    <td>
                      11.08 °C
                    </td>
                    <td>
                      11.08 °C
                    </td>
                  </tr>
                </tbody>
              </table>
            </li>
          </ul>
          <ul>
            <li
              className="tags"
            >
              <button
                onClick={[Function]}
              >
                adver
              </button>
              <button
                onClick={[Function]}
              >
                tise
              </button>
              <button
                onClick={[Function]}
              >
                Ment
              </button>
              <button
                onClick={[Function]}
              >
                339%?
              </button>
            </li>
            <li
              className="tagError"
            >
              
            </li>
            <li>
              <input
                onKeyUp={[Function]}
                placeholder="Add a tag"
                type="text"
              />
            </li>
          </ul>
        </td>
      </tr>
    </tbody>
  </table>
</li>
`;

exports[`testing Weather component with data Matches the snapshot: shows non-expanded Expand component 1`] = `
<li
  className="single_city"
  style={Object {}}
>
  <table>
    <tbody>
      <tr>
        <td>
          <img
            alt="Sukhumi weather icon"
            className="curr_weather"
            src="https://openweathermap.org/img/wn/03d@2x.png"
          />
        </td>
        <td>
          <ul
            className="data_space"
          >
            <li>
              <div
                className="city_name"
                style={
                  Object {
                    "width": "100%",
                  }
                }
              >
                Sukhumi
              </div>
            </li>
            <li>
              <div
                className="country_name"
                style={
                  Object {
                    "width": "100%",
                  }
                }
              >
                Abkhazia
              </div>
            </li>
          </ul>
        </td>
      </tr>
      <tr>
        <td />
        <td>
          <ul
            className="detailWeather"
          >
            <li>
              <button
                className="expand"
                onClick={[Function]}
              >
                +
              </button>
            </li>
            <li>
              <table>
                <thead>
                  <tr>
                    <td>
                      Date & time
                    </td>
                    <td>
                      Current Weather
                    </td>
                    <td>
                      Current Temp
                    </td>
                    <td>
                      Current Body Temp
                    </td>
                    <td>
                      Current Max Temp
                    </td>
                    <td>
                      Current Min Temp
                    </td>
                  </tr>
                </thead>
                <tbody>
                  <tr>
                    <td>
                      undefined
                    </td>
                    <td>
                      <img
                        alt="Sukhumi 2023-01-18 09:00:00 weather icon"
                        className="expand_img"
                        src="https://openweathermap.org/img/wn/03d@2x.png"
                      />
                    </td>
                    <td>
                      11.57 °C
                    </td>
                    <td>
                      10.06 °C
                    </td>
                    <td>
                      11.57 °C
                    </td>
                    <td>
                      14.54 °C
                    </td>
                  </tr>
                </tbody>
              </table>
            </li>
          </ul>
          <ul>
            <li
              className="tags"
            >
              <button
                onClick={[Function]}
              >
                adver
              </button>
              <button
                onClick={[Function]}
              >
                tise
              </button>
              <button
                onClick={[Function]}
              >
                Ment
              </button>
              <button
                onClick={[Function]}
              >
                339%?
              </button>
            </li>
            <li
              className="tagError"
            >
              
            </li>
            <li>
              <input
                onKeyUp={[Function]}
                placeholder="Add a tag"
                type="text"
              />
            </li>
          </ul>
        </td>
      </tr>
    </tbody>
  </table>
</li>
`;
